src/main/java/net/neoforged/neoforge/codec/NBTSerializableCodec.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.codec;
+
+import static net.neoforged.neoforge.codec.NeoforgeNbtCodecs.TAG_CODEC;
+
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import java.util.function.Supplier;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.RegistryOps;
+import net.neoforged.neoforge.common.util.INBTSerializable;
+
+/**
+ * A codec for instances of {@link INBTSerializable}.
+ *
+ * @param defaultSupplier Used to construct empty instances.
+ * @param <S>             Serializer type param.
+ * @param <NBT>           NBT type param.
+ */
+public record NBTSerializableCodec<S extends INBTSerializable<NBT>, NBT extends Tag>(Supplier<S> defaultSupplier) implements Codec<S> {
+    @Override
+    public <T> DataResult<Pair<S, T>> decode(DynamicOps<T> ops, T input) {
+        if (ops instanceof RegistryOps<T> regOps && regOps.lookupProvider instanceof RegistryOps.HolderLookupAdapter adapter) {
```

Comment:
That's a bit weird. Shouldn't we use the RegistryOps.unwrapRegistry or whatever method?